File: code_generation/widget_class.template.py
@widgets.register
class {ClassName}(widgets.DOMWidget):
    """{Description}"""
    _view_name = Unicode('{ClassName}View').tag(sync=True)
    _model_name = Unicode('{ClassName}Model').tag(sync=True)
    _view_module = Unicode('{ModuleName}').tag(sync=True)
    _model_module = Unicode('{ModuleName}').tag(sync=True)
    _view_module_version = Unicode('^{ModuleVersion}').tag(sync=True)
    _model_module_version = Unicode('^{ModuleVersion}').tag(sync=True)

    # python state
{python_state_variables}

    # javascript state
{javascript_state_variables}

    def __init__(self):
        super().__init__()
        self._X = {ClassName}Orig()
        self._X.on_python_state_changed(self._handle_python_state_changed)
        self.observe(self._on_change)
        self._X.init_jupyter()

    def _handle_python_state_changed(self):
        for key in [{python_state_variable_list}]:
            val = self._X.get_python_state(key, None)
            self.set_trait(key, _json_stringify(val))

    def _on_change(self, change):
        if change.type == 'change':
            if change.name in [{javascript_state_variable_list}]:
                state0 = dict()
                state0[change.name] = _json_parse(change.new)
                self._X._handle_javascript_state_changed(state0)

File: code_generation/widget_class.template.js
export class {ClassName}Model extends widgets.DOMWidgetModel {
    defaults = _.extend(widgets.DOMWidgetModel.prototype.defaults(), {
        _model_name : '{ClassName}Model',
        _view_name : '{ClassName}View',
        _model_module : '{ModuleName}',
        _view_module : '{ModuleName}',
        _model_module_version : '{ModuleVersion}',
        _view_module_version : '{ModuleVersion}',

        // Python state
{python_state_defaults}

        // JavaScript state
{javascript_state_defaults}        

    })
}


// Custom View. Renders the widget model.
export class {ClassName}View extends widgets.DOMWidgetView {
    render() {
        this.div=document.createElement('div');
        this.el.appendChild(this.div);

        ReactDOM.render(
            <{ClassName} jupyterModel={this.model} />,
            this.div
        );
    }
}